### ISSUE TITLE
Regression in 3678 - allocation of pending segment fails in some cases 

### ISSUE THREAD

**Issue by pjain1** [druid-io/druid#3741]
After the introduction of https://github.com/druid-io/druid/pull/3678 the Kafka Indexing task can reset offset automatically without supervisor knowing about it. 

Lets say the supervisor assigned some partition to a task to consume and the task reset the offset because the offset was not present in Kafka. It may very well happen that the fetched event falls in the next time interval and not in the same time interval as last pending allocated segment for this sequence. Since, the metadata like lastSegmentIds is restored when the `Appenderator` starts therefore when the task asks overlord to allocate a pending segment for this event, the overlord will get the last pending allocated segment which is not the correct segment and the interval check at https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java#L423 will fail.

I was wondering how necessary is this interval check ? If the interval does not match can we just continue and allocate segment for next interval ? I think the check is present there to prevent from this kind of allocations but is it harmful ? Another option is to pass in some special flag to overlord to indicate that automatic reset has happened and thus don't be so strict about interval checks.

@dclim @gianm

**Comment by gianm** [comment]
@pjain1 do you have replicas or not?

That sanity check is helpful for double checking that replicas are in sync. Two different Kafka tasks that are replicas of each other need to make the _exact same_ decisions based on the Kafka topic, or else replicas can drift and cause inconsistencies. #3678 appears to break this assumption, since it's possible for two replicas to make different decisions about whether and when to reset their offsets. That decision should be made by the supervisor, not by the task. That should be fixed since it means the option from #3678 isn't safe to use with replicas.

You might be running into that or you might be running into something else. If it looks like something else, do you have some more details about what series of events caused the problem you're seeing?

**Comment by pjain1** [comment]
@gianm We do not have any replicas. 

I looked more into it and my previous diagnosis is not exactly correct so please ignore that. Here's what I have observed. 

Task failed to get a new segment id, it did not ingest anything failed on first event (after reset) -

```
2016-12-02T21:59:50,132 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Restored metadata[null].
2016-12-02T21:59:50,136 INFO [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - Seeking partition[24] to offset[12,835,490,176].
...
...
2016-12-02T21:59:50,308 INFO [task-runner-0-priority-0] org.apache.kafka.clients.consumer.internals.Fetcher - Fetch offset 12835490176 is out of range, resetting offset
2016-12-02T21:59:50,309 WARN [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - OffsetOutOfRangeException with message [Offsets out of range with no configured reset policy for partitions: {mx-
cluster-production-metrics-24=12835490176}]
Dec 02, 2016 9:59:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding com.fasterxml.jackson.jaxrs.smile.JacksonSmileProvider to GuiceManagedComponentProvider with the scope "Singleton"
2016-12-02T21:59:50,317 WARN [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - Resetting consumer offset to [latest] for partition [24]
2016-12-02T21:59:50,318 WARN [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - Consumer is now at offset [14090209678]
2016-12-02T21:59:50,744 INFO [task-runner-0-priority-0] io.druid.indexing.common.actions.RemoteTaskActionClient - Performing action for task[index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik]: SegmentAllo
cateAction{dataSource='v2_metrics_cluster', timestamp=2016-12-02T21:59:48.606Z, queryGranularity=DurationGranularity{length=60000, origin=0}, preferredSegmentGranularity=HOUR, sequenceName='index_kafka_v2_metrics_cluster_06ef3c335c14bfd_24', previousSegmentId='null'}
2016-12-02T21:59:50,763 INFO [task-runner-0-priority-0] io.druid.indexing.common.actions.RemoteTaskActionClient - Submitting action for task[index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik] to overlord[/druid/indexer/v1/action]: SegmentAllocateAction{dataSource='v2_metrics_cluster', timestamp=2016-12-02T21:59:48.606Z, queryGranularity=DurationGranularity{length=60000, origin=0}, preferredSegmentGranularity=HOUR, sequenceName='index_kafka_v2_metrics_cluster_06ef3c335c14bfd_24', previousSegmentId='null'}

2016-12-02T21:59:51,701 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[KafkaIndexTask{id=index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik,
 type=index_kafka, dataSource=v2_metrics_cluster}]
io.druid.java.util.common.ISE: Could not allocate segment for row with timestamp[2016-12-02T21:59:48.606Z]
        at io.druid.indexing.kafka.KafkaIndexTask.run(KafkaIndexTask.java:434) ~[druid-kafka-indexing-service-0.9.3-1479913717-13c75d6-1087.jar:0.9.3-1479913717-13c75d6-1087]
        at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:436) [druid-indexing-service-0.9.3-1479913717-13c75d6-1087.jar:0.9.3-1479913717-13c75d6-108
7]
        at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:408) [druid-indexing-service-0.9.3-1479913717-13c75d6-1087.jar:0.9.3-1479913717-13c75d6-108
7]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
2016-12-02T21:59:51,707 INFO [task-runner-0-priority-0] io.druid.indexing.overlord.TaskRunnerUtils - Task [index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik] status changed to [FAILED].
2016-12-02T21:59:51,711 INFO [task-runner-0-priority-0] io.druid.indexing.worker.executor.ExecutorLifecycle - Task completed with status: {
  "id" : "index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik",
  "status" : "FAILED",
  "duration" : 1757
}
```

Overlord logs - 
```
2016-12-02T21:59:51,594 INFO [qtp40910225-116] io.druid.indexing.common.actions.LocalTaskActionClient - Performing action for task[index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik]: SegmentAllocateAction{dataSource='v2_metrics_cluster', timestamp=2016-12-02T21:59:48.606Z, queryGranularity=DurationGranularity{length=60000, origin=0}, preferredSegmentGranularity=HOUR, sequenceName='index_kafka_v2_metrics_cluster_06ef3c335c14bfd_24', previousSegmentId='null'}
2016-12-02T21:59:51,602 INFO [qtp40910225-116] io.druid.indexing.overlord.TaskLockbox - Added task[index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik] to TaskLock[index_kafka_v2_metrics_cluster]
2016-12-02T21:59:51,602 INFO [qtp40910225-116] io.druid.indexing.overlord.MetadataTaskStorage - Adding lock on interval[2016-12-02T21:00:00.000Z/2016-12-02T22:00:00.000Z] version[2016-12-02T21:00:00.244Z] for task: index_kafka_v2_metrics_cluster_06ef3c335c14bfd_elbapfik
2016-12-02T21:59:51,618 WARN [qtp40910225-116] io.druid.metadata.IndexerSQLMetadataStorageCoordinator - Cannot use existing pending segment [v2_metrics_cluster_2016-11-29T17:00:00.000Z_2016-11-29T18:00:00.000Z_2016-11-29T17:00:00.207Z_59] for sequence[index_kafka_v2_metrics_cluster_06ef3c335c14bfd_24] (previous = []) in DB, does not match requested interval[2016-12-02T21:00:00.000Z/2016-12-02T22:00:00.000Z]
```

I can see in the database that there are pending segments in this sequence

```
{
	"data":
	[
		{
			"id": "v2_metrics_cluster_2016-11-29T17:00:00.000Z_2016-11-29T18:00:00.000Z_2016-11-29T17:00:00.207Z_59",
			"dataSource": "v2_metrics_cluster",
			"created_date": "2016-11-29T17:51:40.375Z",
			"start": "2016-11-29T17:00:00.000Z",
			"end": "2016-11-29T18:00:00.000Z",
			"sequence_name": "index_kafka_v2_metrics_cluster_06ef3c335c14bfd_24",
			"sequence_prev_id": "",
			"sequence_name_prev_id_sha1": "1BF32CA26D5F48710C00EF8CDC96F650BA0AF936",
			"payload": "BLOB"
		}
	]
}
```

If the task would have asked to allocate segment with previousSegmentId as `v2_metrics_cluster_2016-11-29T17:00:00.000Z_2016-11-29T18:00:00.000Z_2016-11-29T17:00:00.207Z_59` instead of `null` then it would have been fine and I think that's how it normally works.
